server/cloudspanner_storage_provider.go

```diff
@@ -0,0 +1,117 @@
+	csNumChannels          = flag.Int("cloudspanner_num_channels", 0, "Number of gRPC channels to use to talk to CloudSpanner.")
+	csSessionMaxOpened     = flag.Uint64("cloudspanner_max_open_sessions", 0, "Max open sessions.")
+	csSessionMinOpened     = flag.Uint64("cloudspanner_min_open_sessions", 0, "Min open sessions.")
+	csSessionMaxIdle       = flag.Uint64("cloudspanner_max_idle_sessions", 0, "Max idle sessions.")
+	csSessionMaxBurst      = flag.Uint64("cloudspanner_max_burst_sessions", 0, "Max concurrent create session requests.")
+	csSessionWriteSessions = flag.Float64("cloudspanner_write_sessions", 0, "Fraction of write capable sessions to maintain.")
+	csSessionHCWorkers     = flag.Int("cloudspanner_num_healthcheckers", 0, "Number of health check workers for Spanner session pool.")
+	csSessionHCInterval    = flag.Duration("cloudspanner_healthcheck_interval", 0, "Interval betweek pinging sessions.")
+
+	csMu              sync.RWMutex
+	csStorageInstance *cloudSpannerProvider
+)
+
+func init() {
+	if err := RegisterStorageProvider("cloud_spanner", newCloudSpannerStorageProvider); err != nil {
+		panic(err)
+	}
+}
+
+type cloudSpannerProvider struct {
+	client *spanner.Client
+}
+
+func configFromFlags() spanner.ClientConfig {
+	r := spanner.ClientConfig{}
+	setIntIfNotDefault(&r.NumChannels, *csNumChannels)
+	setUint64IfNotDefault(&r.SessionPoolConfig.MaxOpened, *csSessionMaxOpened)
+	setUint64IfNotDefault(&r.SessionPoolConfig.MinOpened, *csSessionMinOpened)
+	setUint64IfNotDefault(&r.SessionPoolConfig.MaxIdle, *csSessionMaxIdle)
+	setUint64IfNotDefault(&r.SessionPoolConfig.MaxBurst, *csSessionMaxBurst)
+	setFloat64IfNotDefault(&r.SessionPoolConfig.WriteSessions, *csSessionWriteSessions)
+	setIntIfNotDefault(&r.SessionPoolConfig.HealthCheckWorkers, *csSessionHCWorkers)
+	r.SessionPoolConfig.HealthCheckInterval = *csSessionHCInterval
+	return r
+}
+
+func newCloudSpannerStorageProvider(mf monitoring.MetricFactory) (StorageProvider, error) {
+	csMu.Lock()
+	defer csMu.Unlock()
+
+	if csStorageInstance != nil {
+		return csStorageInstance, nil
+	}
+
+	client, err := spanner.NewClientWithConfig(context.TODO(), *csURI, configFromFlags())
+	if err != nil {
+		return nil, err
+	}
+	csStorageInstance = &cloudSpannerProvider{
+		client: client,
+	}
+	return csStorageInstance, nil
+}
+
+func (s *cloudSpannerProvider) LogStorage() storage.LogStorage {
+	return cloudspanner.NewLogStorage(s.client)
+}
+
+func (s *cloudSpannerProvider) MapStorage() storage.MapStorage {
+	return nil
+}
+
+func (s *cloudSpannerProvider) AdminStorage() storage.AdminStorage {
+	return cloudspanner.NewAdminStorage(s.client)
+}
+
+func (s *cloudSpannerProvider) Close() error {
+	s.client.Close()
+	return nil
+}
+
+func setIntIfNotDefault(t *int, v int) {
+	if v != 0 {
+		*t = v
+	}
+}
+
+func setUint64IfNotDefault(t *uint64, v uint64) {
+	if v != 0 {
+		*t = v
+	}
+}
+
+func setFloat64IfNotDefault(t *float64, v float64) {
+	if v != 0 {
```

Comment:
Unsafe float comparison? Not sure if zero is ok for this.

Reply:
golang doesn't allow you to compare dissimilar types, and since this zero is an untyped constant I believe the language spec requires the compiler to define it as a `float64` with zero value (because of the expression it's part of) - same as the default values for the flags.

Reply:
What I wrote might not have been clear. I meant that it's generally not a good idea to compare floats for equality. Zero might be a special case though but I haven't looked at any specs.

Reply:
golang uses IEEE754 which has an explicit representation for zero, and golang language spec says:
"Numeric constants represent exact values of arbitrary precision and do not overflow. Consequently, there are no constants denoting the IEEE-754 negative zero, infinity, and not-a-number values."

So I think this is pretty safe.